Mathlib/Data/Fin/Basic.lean

```diff
@@ -1,334 +1,2533 @@
+    a.succ.predAbove b.succ = (a.predAbove b).succ := by
+  obtain h₁ | h₂ := lt_or_le (castSucc a) b
+  · rw [Fin.predAbove_above _ _ h₁, Fin.succ_pred, Fin.predAbove_above, Fin.pred_succ]
+    simpa only [lt_iff_val_lt_val, coe_castSucc, val_succ, add_lt_add_iff_right] using
+      h₁
+  · cases' n with n
+    · exfalso
+      exact not_lt_zero' a.is_lt
+    · rw [Fin.predAbove_below a b h₂,
+        Fin.predAbove_below a.succ b.succ
+          (by
+            simpa only [le_iff_val_le_val, val_succ, coe_castSucc, add_le_add_iff_right] using h₂)]
+      ext
+      have h₀ : (b : ℕ) < n + 1 := by
+        simp only [le_iff_val_le_val, coe_castSucc] at h₂
+        simpa only [lt_succ_iff] using h₂.trans a.is_le
+      have h₁ : (b.succ : ℕ) < n + 2 := by
+        rw [← Nat.succ_lt_succ_iff] at h₀
+        simpa only [val_succ] using h₀
+      simp only [coe_castPred b h₀, coe_castPred b.succ h₁, val_succ]
+#align fin.succ_pred_above_succ Fin.succ_predAbove_succ
+
+@[simp]
+theorem castPred_castSucc (i : Fin (n + 1)) : castPred (castSucc i) = i := by
+  simp [castPred, predAbove, not_lt.mpr (le_last i)]
+#align fin.cast_pred_cast_succ Fin.castPred_castSucc
+
+theorem castSucc_castPred {i : Fin (n + 2)} (h : i < last (n + 1)) : castSucc i.castPred = i := by
+  rw [castPred, predAbove, dif_neg]
+  · simp [Fin.eq_iff_veq]
+  · exact h.not_le
+#align fin.cast_succ_cast_pred Fin.castSucc_castPred
+
+theorem coe_castPred_le_self (i : Fin (n + 2)) : (i.castPred : ℕ) ≤ i := by
+  rcases i.le_last.eq_or_lt with (rfl | h)
+  · simp
+  · rw [castPred, predAbove, dif_neg]
+    · simp
+    · simpa [lt_iff_val_lt_val, le_iff_val_le_val, lt_succ_iff] using h
+#align fin.coe_cast_pred_le_self Fin.coe_castPred_le_self
+
+theorem coe_castPred_lt_iff {i : Fin (n + 2)} : (i.castPred : ℕ) < i ↔ i = Fin.last _ := by
+  rcases i.le_last.eq_or_lt with (rfl | H)
+  · simp
+  · simp only [_root_.ne_of_lt H]
+    rw [← castSucc_castPred H]
+    simp
+#align fin.coe_cast_pred_lt_iff Fin.coe_castPred_lt_iff
+
+theorem lt_last_iff_coe_castPred {i : Fin (n + 2)} :
+    i < Fin.last _ ↔ (i.castPred : ℕ) = i := by
+  rcases i.le_last.eq_or_lt with (rfl | H)
+  · simp
+  · simp only [H]
+    rw [← castSucc_castPred H]
+    simp
+#align fin.lt_last_iff_coe_cast_pred Fin.lt_last_iff_coe_castPred
+
+end PredAbove
+
+/-- `min n m` as an element of `Fin (m + 1)` -/
+def clamp (n m : ℕ) : Fin (m + 1) :=
+  OfNat.ofNat <| min n m
+#align fin.clamp Fin.clamp
+
+@[simp]
+theorem coe_clamp (n m : ℕ) : (clamp n m : ℕ) = min n m :=
+  Nat.mod_eq_of_lt <| Nat.lt_succ_iff.mpr <| min_le_right _ _
+#align fin.coe_clamp Fin.coe_clamp
+
+@[simp]
+theorem coe_ofNat_eq_mod' (m n : ℕ) [NeZero m] :
+    (@Fin.ofNat' m n (Nat.pos_of_ne_zero (NeZero.ne m)) : ℕ) = n % m :=
+  rfl
+#align fin.coe_of_nat_eq_mod' Fin.coe_ofNat_eq_mod'
+
+@[simp]
+theorem coe_of_nat_eq_mod (m n : ℕ) : ((n : Fin (m+1)) : ℕ) = n % Nat.succ m :=
+  rfl
+#align fin.coe_of_nat_eq_mod Fin.coe_of_nat_eq_mod
```

Comment:
seems we need a matching PR to your [#18131](https://github.com/leanprover-community/mathlib/pull/18131). @negiizhao will you do it?

Reply:
Just noticed the LHS of these two theorems look syntactically different, but they were the same in mathlib3. The latter is now removed from mathlib3, and the former is renamed to the latter, so the first #align should fail and the second #align should be a misalignment now. Can we change the statement of the former to `((n : Fin m) : ℕ) = n % m` and remove the latter?